mobile_config/src/gateway_service.rs

```diff
@@ -191,27 +198,39 @@ impl mobile_config::Gateway for GatewayService {
         self.verify_request_signature(&signer, &request)?;
 
         let pool = self.metadata_pool.clone();
+        let mc_pool = self.mobile_config_db_pool.clone();
         let signing_key = self.signing_key.clone();
         let batch_size = request.batch_size;
 
         let (tx, rx) = tokio::sync::mpsc::channel(100);
 
         let device_types: Vec<DeviceType> = request.device_types().map(|v| v.into()).collect();
-        let min_refreshed_at = Utc
-            .timestamp_opt(request.min_refreshed_at as i64, 0)
-            .single()
-            .ok_or(Status::invalid_argument(
-                "Invalid min_refreshed_at argument",
-            ))?;
 
         tracing::debug!(
             "fetching all gateways' info (v2). Device types: {:?} ",
             device_types
         );
 
         tokio::spawn(async move {
-            let stream = gateway_info::db::all_info_stream(&pool, &device_types, min_refreshed_at);
-            stream_multi_gateways_info(stream, tx.clone(), signing_key.clone(), batch_size).await
+            let stream = gateway_info::db::all_info_stream(&pool, &device_types);
+            if request.min_updated_at > 0 {
+                let min_updated_at = Utc
+                    .timestamp_opt(request.min_updated_at as i64, 0)
+                    .single()
+                    .ok_or(Status::invalid_argument(
+                        "Invalid min_refreshed_at argument",
+                    ))?;
+
+                let updated_redios = get_updated_radios(&mc_pool, min_updated_at).await?;
+                let stream = stream
+                    .filter(|v| future::ready(updated_redios.contains(&v.address)))
+                    .boxed();
+                stream_multi_gateways_info(stream, tx.clone(), signing_key.clone(), batch_size)
+                    .await
+            } else {
+                stream_multi_gateways_info(stream, tx.clone(), signing_key.clone(), batch_size)
+                    .await
+            }
```

Comment:
```suggestion
            let stream =
                if request.min_updated_at > 0 {
                    let min_updated_at = Utc
                        .timestamp_opt(request.min_updated_at as i64, 0)
                        .single()
                        .ok_or(Status::invalid_argument(
                            "Invalid min_refreshed_at argument",
                        ))?;

                    let updated_redios = get_updated_radios(&mc_pool, min_updated_at).await?;
                    gateway_info::db::all_info_stream(&pool, &device_types)
                        .filter(|v| future::ready(updated_redios.contains(&v.address)))
                        .boxed();
                } else {
                    gateway_info::db::all_info_stream(&pool, &device_types)
                };
            stream_multi_gateways_info(stream, tx.clone(), signing_key.clone(), batch_size).await
```
i don't know that this really makes a difference; i wanted to get it down to a single invocation of the stream but i don't think we can fully clean that up unless we added a no-op filter on the branch that _doesn't_ check updated at so feel free to ignore here

Reply:
I've tried to do it in a similar way, but I'm facing incompatible types or borrow checker errors.